aya-log-common/src/lib.rs

```diff
@@ -96,55 +99,62 @@ pub enum DisplayHint {
     UpperMac,
 }
 
-struct TagLenValue<'a, T> {
-    tag: T,
-    value: &'a [u8],
+struct TagLenValue<T, V> {
+    pub tag: T,
+    pub value: V,
 }
 
-impl<'a, T> TagLenValue<'a, T>
+impl<T, V> TagLenValue<T, V>
 where
-    T: Copy,
+    V: IntoIterator<Item = u8>,
+    <V as IntoIterator>::IntoIter: ExactSizeIterator,
 {
-    #[inline(always)]
-    pub(crate) fn new(tag: T, value: &'a [u8]) -> TagLenValue<'a, T> {
-        TagLenValue { tag, value }
-    }
-
-    pub(crate) fn write(&self, mut buf: &mut [u8]) -> Result<usize, ()> {
-        let size = mem::size_of::<T>() + mem::size_of::<usize>() + self.value.len();
-        let remaining = cmp::min(buf.len(), LOG_BUF_CAPACITY);
+    pub(crate) fn write(self, mut buf: &mut [u8]) -> Result<usize, ()> {
+        let Self { tag, value } = self;
+        let value = value.into_iter();
+        let len = value.len();
+        let wire_len: LogValueLength = value
+            .len()
+            .try_into()
+            .map_err(|num::TryFromIntError { .. }| ())?;
+        let size = mem::size_of_val(&tag) + mem::size_of_val(&wire_len) + len;
         // Check if the size doesn't exceed the buffer bounds.
-        if size > remaining {
+        if size > buf.len() {
```

Comment:
I removed it because I ran this and the verifier was happy. Want to test it on your setup and let me know if the verifier gets mad?

Reply:
I think we should put it back. We can't guarantee that LLVM is going to codegen in a way that the verifier is always going to be able to statically determine the value of buf.len(). This is a common issue with slices, and I remember hitting this issue in this particular point of the code.

Reply:
What you describe works for network direct packet access, where the verifier has two special registers data_start and data_end that mark the packet boundaries and that can be used to dynamically do bound checking.

We're writing log records to a map, which doesn't have special marker registers. The only thing the verifier knows is the fixed capacity of the map, which it will enforce statically while it runs state tracking (simulates all possible code paths). The buf.len() check will only work if the slice is actually optimized out at compile time, and so buf.len() is always a register holding a constant. Since we can't guarantee that, we need to clamp with an actual constant.

Reply:
Done.